src/crab/ebpf_domain.cpp

```diff
@@ -1166,134 +1387,172 @@ void ebpf_domain_t::operator()(const Bin& bin) {
         }
     } else {
         // dst op= src
-        auto src = reg_pack(std::get<Reg>(bin.v));
+        auto src_reg = std::get<Reg>(bin.v);
+        auto src = reg_pack(src_reg);
         switch (bin.op) {
         case Bin::Op::ADD: {
             if (type_inv.same_type(m_inv, bin.dst, std::get<Reg>(bin.v))) {
                 // both must be numbers
                 add_overflow(dst.value, src.value);
             } else {
                 // Here we're not sure that lhs and rhs are the same type; they might be.
-                // But previous assertions should fail unless we know that exactly one of lhs or rhs to is a pointer
-                m_inv = type_inv.join_by_if_else(m_inv,
-                    type_is_number(bin.dst),
-                    [&](NumAbsDomain& inv) {
-                        // num + ptr
-                        apply(inv, crab::arith_binop_t::ADD, dst.offset, dst.value, src.offset, false);
-                        type_inv.assign_type(inv, bin.dst, std::get<Reg>(bin.v));
-                        inv.assign(dst.region_size, src.region_size);
-                    },
-                    [&](NumAbsDomain& inv) {
-                        // ptr + num
-                        apply(inv, crab::arith_binop_t::ADD, dst.offset, dst.offset, src.value, false);
-                    }
-                );
+                // But previous assertions should fail unless we know that exactly one of lhs or rhs is a pointer.
+                m_inv = type_inv.join_over_types(m_inv, bin.dst, [&](NumAbsDomain& inv, type_encoding_t dst_type) {
+                    inv = type_inv.join_over_types(m_inv, src_reg, [&](NumAbsDomain& inv, type_encoding_t src_type) {
+                        if (dst_type == T_NUM && src_type != T_NUM) {
+                            // num += ptr
+                            type_inv.assign_type(inv, bin.dst, src_type);
+                            auto dst_offset = get_type_offset_variable(bin.dst, src_type);
+                            if (dst_offset.has_value())
+                                apply(inv, crab::arith_binop_t::ADD, dst_offset.value(), dst.value,
+                                      get_type_offset_variable(src_reg, src_type).value(), false);
+                            if (src_type == T_SHARED)
+                                inv.assign(dst.shared_region_size, src.shared_region_size);
+                        } else if (dst_type != T_NUM && src_type == T_NUM) {
+                            // ptr += num
+                            type_inv.assign_type(inv, bin.dst, dst_type);
+                            auto dst_offset = get_type_offset_variable(bin.dst, dst_type);
+                            if (dst_offset.has_value())
+                                apply(inv, crab::arith_binop_t::ADD, dst_offset.value(), dst_offset.value(), src.value,
+                                      false);
+                        } else {
+                            inv.set_to_bottom();
```

Comment:
I think this is correct, but worth an explanation in comment. Basically we ignore here the cases that do not match the assumption, as described in the comment just before the join_over_types. Another possibility would be to set to top, which is easier to reason about but means we'll get more warnings later (since everything will become top), and that will be less helpful.

Reply:
Added comment